src/lib.rs

```diff
@@ -14,39 +24,184 @@ fn process_instruction(
+            invoke_signed(
+                &system_instruction::create_account(
+                    funder_info.key,
+                    program_token_info.key,
+                    1.max(rent.minimum_balance(spl_token::state::Account::get_packed_len())),
+                    spl_token::state::Account::get_packed_len() as u64,
+                    &spl_token::id(),
+                ),
+                &[
+                    funder_info.clone(),
+                    program_token_info.clone(),
+                    system_program_info.clone(),
+                ],
+                &[&program_token_signer_seeds],
+            )?;
+
+            info!("Initializing program token account");
+            invoke(
+                &spl_token::instruction::initialize_account(
+                    &spl_token::id(),
+                    program_token_info.key,
+                    mint_info.key,
+                    program_token_info.key, // token owner is also `program_token` address
+                )?,
+                &[
+                    program_token_info.clone(),
+                    spl_token_program_info.clone(),
+                    rent_sysvar_info.clone(),
+                    mint_info.clone(),
+                ],
+            )?;
+            Ok(())
+        }
+        Some(1) => {
+            info!("Close program token account...");
+            let funder_info = next_account_info(account_info_iter)?;
+            let spl_token_program_info = next_account_info(account_info_iter)?;
+
+            invoke_signed(
+                &spl_token::instruction::close_account(
+                    &spl_token::id(),
+                    program_token_info.key,
+                    funder_info.key,
+                    program_token_info.key, // token owner is also `program_token` address
+                    &[],
+                )
+                .expect("close_account"),
+                &[
+                    funder_info.clone(),
+                    spl_token_program_info.clone(),
+                    program_token_info.clone(),
+                ],
+                &[&program_token_signer_seeds],
+            )
+        }
+        _ => {
+            info!("Error: Unsupported instruction");
+            Err(ProgramError::InvalidInstructionData)
+        }
+    }
 }
 
 #[cfg(test)]
 mod test {
+    #![cfg(feature = "test-bpf")]
+
     use super::*;
     use assert_matches::*;
-    use solana_program::instruction::{AccountMeta, Instruction};
+    use solana_program::{
+        instruction::{AccountMeta, Instruction},
+        sysvar,
+    };
     use solana_program_test::*;
     use solana_sdk::{signature::Signer, transaction::Transaction};
 
+    fn program_test(program_id: Pubkey) -> ProgramTest {
+        let mut pc = ProgramTest::new(
+            "bpf_program_template",
+            program_id,
+            processor!(process_instruction),
+        );
+
+        // Add SPL Token program
+        pc.add_program(
+            "spl_token",
```

Comment:
should this be `program-token` to match line 148? or is this not meant to match the seed?

Reply:
i ask because a new `program_id` pubkey is created on line 145, and i don't see any other way the test would identify the `spl-token` address

Reply:
No, this `"spl_token"` string is used within program test itself to potentially load the SPL Token program from the filesystem during test execution (code for that is here FYI: https://github.com/solana-labs/solana/blob/89b474e192ea2f4b2274ad26cbc47443d5767571/program-test/src/lib.rs#L480-L487)

> i ask because a new program_id pubkey is created on line 145

This just generates a placeholder address for the `bpf_program_template` program for the purposes of the test.  In a real program deployment, `solana deploy` would generate this address for you if you didn't explicitly provide a deploy keypair (the 2nd argument to `solana deploy`)

>  don't see any other way the test would identify the spl-token address

Line 148 is where the test figures the spl-token address used by the program:
```rust
let program_token_address =
            Pubkey::find_program_address(&[br"program-token"], &program_id).0;
```

Reply:
it seems the `spl_token` is just an arbitrary name, not related to anything else in the program so disregard the above comments

Reply:
ha, didn't see this, had commented in another window